### PR TITLE
Rewrite cartridge mock for unit tests

### PR DESCRIPTION
**Pull request by dokshina**
Before this patch `cartridge` mock used for unit tests emulated
`cartridge` logic with mocked functions. But it became complicated 
to support `cartridge`-like logic. Since this patch real `cartridge`  
functions are called, but some networking logic is mocked. All called 
functions are wrapped to save argument calls (to check them in tests)
or return an error when it is necessary for tests.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,6 +1,7 @@
 ### Project ###
 
 packages/
+**/.rocks
```
Comment (hackallcode) [inline .gitignore]: It works on any subdirectory
```suggestion
.rocks
```

File: .github/workflows/unit-tests.yml [hunk 2]
```diff
@@ -32,8 +32,21 @@ jobs:
...
+      - name: Cache unit tests requirements
+        id: cache-packages
+        uses: actions/cache@v2
+        with:
+          path: 'unit/mock/.rocks'
+          key: ce-${{ matrix.tarantool-version }}-${{ hashFiles('./deps.sh') }}
+
+      - name: Install unit tests requirements
+        run: |
+          ./deps.sh
```
Comment (hackallcode) [inline .github/workflows/unit-tests.yml]:
```suggestion
        run: ./deps.sh
```

File: deps.sh [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+# Call this script to install unit test dependencies
+
+set -e
+
+pushd ./unit/mock
+
+tarantoolctl rocks install cartridge 2.4.0
```
Comment (hackallcode) [inline deps.sh:8]: What about to use version from environment or argument to specify cartridge version?

File: library/cartridge_edit_topology.py [hunk 4]
```diff
@@ -284,7 +284,7 @@ def get_edit_servers_params(instances_to_expel, cluster_instances):
             'expelled': True,
         }
         for instance_name in instances_to_expel
-        if instance_name in cluster_instances and cluster_instances[instance_name].get('uuid') is not None
+        if instance_name in cluster_instances and cluster_instances[instance_name].get('uuid')
```
Comment (hackallcode) [inline library/cartridge_edit_topology.py:287]: Is it possible that `uuid == ''`? Or for what this patch?

File: unit/instance.py [hunk 5]
```diff
@@ -8,7 +8,7 @@
 
 from unit.os_mock import OsPathExistsMock, OsPathGetMTimeMock
 
-script_abspath = os.path.realpath(
+instance_path = os.path.realpath(
```
Comment (hackallcode) [inline unit/instance.py]: `mock_path`? `instance_app_dir`?

File: unit/instance.py [hunk 6]
```diff
@@ -29,10 +29,11 @@ class Instance:
     DATE_TODAY = 0
     DATE_TOMORROW = 1
 
-    def __init__(self, console_sock, cluster_cookie):
+    def __init__(self):
+        self.console_sock = os.path.join(os.getcwd(), 'x.sock')
```
Comment (hackallcode) [inline unit/instance.py]: `x.sock` is a strange name

File: unit/instance.py [hunk 7]
```diff
@@ -57,13 +60,18 @@ def _connect(self):
...
         env = os.environ.copy()
         env['TARANTOOL_CONSOLE_SOCK'] = self.console_sock
 
         with open(os.devnull, 'w') as FNULL:
-            self.process = Popen(command, env=env, stdout=FNULL, stderr=FNULL)
+            self.process = Popen(command, env=env, cwd=cwd, stdout=FNULL, stderr=FNULL)
+
+        # # verbose output
+        # import sys
+        # self.process = Popen(command, env=env, cwd=cwd, stdout=sys.stdout, stderr=sys.stderr)
```
Comment (hackallcode) [inline unit/instance.py]: Debug

File: unit/instance.py [hunk 8]
```diff
@@ -88,28 +96,33 @@ def recvall():
                 chunk = self.sock.recv(1024).decode()
                 # It is correct because of cmd structure: it always returns a value
                 if chunk == '':
-                    raise Exception(
-                        'Error: broken pipe. '
-                        'Probably, the instance was not bootstrapped yet to perform this operation'
-                    )
+                    errmsg = 'Error: broken pipe. ' + \
+                             'Probably, the instance was not bootstrapped yet to perform this operation'
+                    raise Exception(errmsg)
```
Comment (hackallcode) [inline unit/instance.py]: A strange change)

File: unit/instance.py [hunk 9]
```diff
@@ -88,28 +96,33 @@ def recvall():
...
+end
+local args = require('json').decode('{args_encoded}')
+local ret = {{
+    load(
+        'local func, args = ... return func(unpack(args))',
+        '@eval'
+    )(func, args)
+}}
+return string.hex(require('json').encode(ret))
+'''
```
Comment (hackallcode) [inline unit/instance.py]:
```suggestion
        cmd_fmt = textwrap.dedent('''
            local function func(...)
                {func_body}
            end
            local args = require('json').decode('{args_encoded}')
            local ret = {{
                load(
                    'local func, args = ... return func(unpack(args))',
                    '@eval'
                )(func, args)
            }}
            return string.hex(require('json').encode(ret))
        ''')
```

File: unit/instance.py [hunk 10]
```diff
@@ -122,17 +135,35 @@ def recvall():
...
-        except Exception:
-            raise Exception(hex_output)
+        if hex_output.startswith("error:"):
+            err = re.sub(r"error:\s+", '', hex_output)
+            raise Exception(err)
+
+        output = bytearray.fromhex(hex_output).decode('utf-8')
+
+        data = json.loads(output)
+        return data
```
Comment (hackallcode) [inline unit/instance.py]:
```suggestion
        return json.loads(output)
```

File: unit/instance.py [hunk 11]
```diff
@@ -201,129 +233,75 @@ def set_path_m_time(path, m_time):
...
+    def add_replicaset(self, **kwargs):
+        self.eval_res_err('''
+            require('cartridge').internal.add_replicaset(...)
+        ''', kwargs)
+
+    def bootstrap_cluster(self):
+        res, err = self.eval_res_err('''
+            require('cartridge').internal.bootstrap_cluster()
+        ''')
+        assert err is None, err
```
Comment (hackallcode) [inline unit/instance.py]: What about to check errors in other places? Or maybe return an error?

File: unit/mock/init.lua [hunk 12]
```diff
@@ -2,9 +2,17 @@
 require('strict').on()
 
 local fio = require('fio')
+local app_dir = fio.abspath(fio.dirname(arg[0]))
 
-local script_dir = fio.abspath(fio.dirname(arg[0]))
-package.path = package.path .. ';' .. script_dir .. '/?.lua'
+print('App dir set to ' .. app_dir)
```
Comment (hackallcode) [inline unit/mock/init.lua]: Debug?

File: library/cartridge_edit_topology.py (hunk 4 above)
Comment (dokshina) [reply library/cartridge_edit_topology.py:287]: Yes, it turned out that `cartridge` returns `uuid = ''` for unjoined instances.

File: unit/instance.py (hunk 6 above)
Comment (dokshina) [reply unit/instance.py]: Renamed to `instance.sock`.

File: unit/instance.py (hunk 8 above)
Comment (dokshina) [reply unit/instance.py]: That was copy paste.

File: unit/instance.py (hunk 11 above)
Comment (dokshina) [reply unit/instance.py]: I've added asserts in Lua code for internal functions. `eval_res_err` will raise exceptions for this errors with assert message.

File: unit/instance.py (hunk 7 above)
Comment (dokshina) [reply unit/instance.py]: I've added `debug` flag to `instance.start`.

File: unit/instance.py [hunk 13]
```diff
@@ -122,17 +140,34 @@ def recvall():
...
-        return ret['ret']
+        if len(data) == 0:
+            # return nil
+            data.append(None)
+
+        if len(data) == 1:
+            # return res
+            data.append(None)
+
+        # return res, err
```
Comment (hackallcode) [inline unit/instance.py]: Remove it

**Review by hackallcode — APPROVED** [review]
In the near future, it will be necessary to divide these tests into unit tests and integration tests.

**Pull request merged**